PhysicsTools/NanoAOD/plugins/IsoValueMapProducer.cc

```diff
@@ -280,21 +362,39 @@ void IsoValueMapProducer<T>::fillDescriptions(edm::ConfigurationDescriptions& de
         ->setComment("rho to be used for effective-area based PF-isolation pileup subtraction for electrons");
   }
   if ((typeid(T) == typeid(pat::Photon))) {
-    desc.add<edm::InputTag>("mapIsoChg")->setComment("input charged PF isolation calculated in VID for photons");
-    desc.add<edm::InputTag>("mapIsoNeu")->setComment("input neutral PF isolation calculated in VID for photons");
-    desc.add<edm::InputTag>("mapIsoPho")->setComment("input photon PF isolation calculated in VID for photons");
-    desc.add<edm::FileInPath>("EAFile_PFIso_Chg")
+    desc.addOptional<edm::InputTag>("mapIsoChg")
+        ->setComment("input charged PF isolation calculated in VID for photons");
+    desc.addOptional<edm::InputTag>("mapIsoNeu")
+        ->setComment("input neutral PF isolation calculated in VID for photons");
+    desc.addOptional<edm::InputTag>("mapIsoPho")->setComment("input photon PF isolation calculated in VID for photons");
+
+    desc.addOptional<edm::FileInPath>("EAFile_PFIso_Chg")
         ->setComment(
             "txt file containing effective areas to be used for charged PF-isolation pileup subtraction for photons");
-    desc.add<edm::FileInPath>("EAFile_PFIso_Neu")
+    desc.addOptional<edm::FileInPath>("EAFile_PFIso_Neu")
         ->setComment(
             "txt file containing effective areas to be used for neutral PF-isolation pileup subtraction for photons");
-    desc.add<edm::FileInPath>("EAFile_PFIso_Pho")
+    desc.addOptional<edm::FileInPath>("EAFile_PFIso_Pho")
         ->setComment(
             "txt file containing effective areas to be used for photon PF-isolation pileup subtraction for photons");
+
     desc.add<edm::InputTag>("rho_PFIso")
```

Comment:
I'm curious how do you decide to make a parameter "optional" or not. For example `rho_PFIso` is only used by electrons and photons: why it is not set optional as other parameters not included in all configurations?

There are other such parameters only used for specific particles, which are not set optional: please check that all that is correct

Reply:
The effective Area (EA) correction for `Fall17V2` has different definition (compared to `RunIIIWinter22V1`) and consequently different EA files (.txt) where the values are stored. `QuadraticEAFile_PFIso_` are used only in `RunIIIWinter22V1` while `EAFile_PFIso_` are used only in `Fall17V2`. And that is why parameters corresponding to them are kept optional because only one kind of EA correction is accessed by a particular era of ID. The `mapIso` parameters are neither used in the plugin nor in the photon table. WF like 2500.6 accesses the `V10/nano_cff.py` which has these parameters in its config. Since, we are not touching the V10, we made the `mapIso` parameters optional as well, while not having these parameters in the current `nano_cff.py` config. If I don't do this, WF 2500.6 crashes. `rho_PFIso` is used by both (`Fall17V2` and `RunIIIWinter22V1`) the IDs and hence, it is kept as 'non-optional'.

Reply:
@shdutta16 Deciding whether a parameter is Optional or not shouldn't depend on whether it is already present in the existing python configurations. It should rather depend on the structure of the code.

As a possible rule: a parameter should be considered optional if a different behaviour in the code depends on whether the parameter exists or not, or if it is technically possible that such a parameter is not needed in the execution flow (for example if it is only relevant for a given case, e.g. a particle, accessible depending on the value assigned to another configuration parameter). If when a parameter is missing a default value is assigned to it, and the parameter itself with that default value continues to be used in the flow of the code, it is not an optional parameter any more.

If you have different parameters for different particles, I'd say that they should all be considered "optional", unless you specify a default value for them and use it when the parameter is not explicitely listed in the python configuration.

Reply:
If we understand your comment correctly, that is what we are trying to do :

1. The parameters that are optional based on which era of ID is calling `IsoValueMapProducer`, have been kept optional. `QuadraticEAFile_PFIso_` is called by `RunIIIWinter22V1` ID and `EAFile_PFIso_` is called by `Fall17V2` ID. 

2. `rho_PFIso` is essential for Photon and we want the code to throw error if it is not present, and that's why it is not kept optional. 

3. `mapIso` doesn't have any usage in the Run3 scenario, but for nanoAOD V10 they are needed and in a previous comment here, we were advised not to touch V10. So, they are kept optional as well, otherwise V10 workflow will crash which has already been tested.

Reply:
Uhm, probably the logic behind is more convoluted than I can imagine from my glances to the code.
Taking into account that those parameters were already as such before your PR, let avoid touching them here.
However, I'm convinced that the whole parameter structure could be simplified, to the ease of future developers and maintainers.
Let merge this as it is now, then.